golang/cosmos/types/kv_entry.go

```diff
@@ -0,0 +1,111 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+var _ json.Marshaler = &KVEntry{}
+var _ json.Unmarshaler = &KVEntry{}
+
+// KVEntry represents a string key / string value pair, where the value may be
+// missing. The semantics of a missing value are up to the user.
+type KVEntry struct {
+	key   string
+	value *string
+}
+
+// NewKVEntry creates a KVEntry with the provided key and value
+func NewKVEntry(key string, value string) KVEntry {
+	return KVEntry{key, &value}
+}
+
+// NewKVEntryWithNoValue creates a KVEntry with the provided key and no value
+func NewKVEntryWithNoValue(key string) KVEntry {
+	return KVEntry{key, nil}
+}
+
+// UnmarshalJSON interprets the data as a [key: string, value?: string | null]
+// JSON array and updated the current KVEntry.
+// The key must be a string, and the value (if present) must be a string or null.
+//
+// Implements json.Unmarshaler
+// Note: unlike other methods, this accepts a pointer to satisfy
+// the Unmarshaler semantics.
+func (entry *KVEntry) UnmarshalJSON(msg []byte) (err error) {
+	var generic []interface{}
+	err = json.Unmarshal(msg, &generic)
+
+	if err != nil {
+		return err
+	}
```

Comment:
Hmm, I don't think errors like "_json: cannot unmarshal … into Go value of type []interface {} []_" are useful here (unlike those reporting details of invalid JSON).
```suggestion
	var unmarshaled any
	if err = json.Unmarshal(msg, &unmarshaled); err != nil {
		return err
	}
	generic, ok := unmarshaled.([]any)
	if !ok {
		return fmt.Errorf("KVEntry must be an array")
	}
```

Reply:
It's an interesting suggestion but I think it's incompatible with the suggestion @JimLarson made to use `[]*string`. I might be ok with the JSON marshall errors pop through

Reply:
I would prefer more specific errors, but "_json: cannot unmarshal … into Go value of type string_" isn't terrible. :+1: